tests/server/client.go

```diff
@@ -0,0 +1,103 @@
+package testsuite
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	abcicli "github.com/tendermint/abci/client"
+	"github.com/tendermint/abci/types"
+	crypto "github.com/tendermint/go-crypto"
+	cmn "github.com/tendermint/tmlibs/common"
+)
+
+func InitChain(client abcicli.Client) error {
+	total := 10
+	vals := make([]*types.Validator, total)
+	for i := 0; i < total; i++ {
+		pubkey := crypto.GenPrivKeyEd25519FromSecret([]byte(cmn.Fmt("test%d", i))).PubKey().Bytes()
+		power := cmn.RandInt()
+		vals[i] = &types.Validator{pubkey, uint64(power)}
+	}
+	err := client.InitChainSync(types.RequestInitChain{Validators: vals})
+	if err != nil {
+		fmt.Println("Failed test: InitChain - %v", err)
+		return err
+	}
+	fmt.Println("Passed test: InitChain")
+	return nil
+}
+
+func SetOption(client abcicli.Client, key, value string) error {
+	res := client.SetOptionSync(key, value)
+	_, _, log := res.Code, res.Data, res.Log
+	if res.IsErr() {
+		fmt.Println("Failed test: SetOption")
+		fmt.Printf("setting %v=%v: \nlog: %v", key, value, log)
+		fmt.Println("Failed test: SetOption")
+		return errors.New(res.Error())
+	}
+	fmt.Println("Passed test: SetOption")
+	return nil
+}
+
+func Commit(client abcicli.Client, hashExp []byte) error {
+	res := client.CommitSync()
+	_, data, log := res.Code, res.Data, res.Log
+	if res.IsErr() {
+		fmt.Println("Failed test: Commit")
+		fmt.Printf("committing %v\nlog: %v", log)
+		return errors.New(res.Error())
+	}
+	if !bytes.Equal(data, hashExp) {
+		fmt.Println("Failed test: Commit")
+		fmt.Printf("Commit hash was unexpected. Got %X expected %X",
+			data, hashExp)
+		return errors.New("CommitTx failed")
+	}
+	fmt.Println("Passed test: Commit")
+	return nil
+}
+
+func DeliverTx(client abcicli.Client, txBytes []byte, codeExp types.CodeType, dataExp []byte) error {
+	res := client.DeliverTxSync(txBytes)
```

Comment:
what about checking the `res.IsErr()`? similar to `CheckTx`

Reply:
Looks like `res.IsErr()` checks whether code is not `OK`. However we already have the `codeExp` argument. I guess that the check of `res.IsErr()` should be removed from `CheckTx` then, as it is redundant and will lead to mismatch if the caller expects something different than `OK`.